package/MDAnalysis/analysis/diffusionmap.py

```diff
@@ -0,0 +1,419 @@
+            Any valid selection string for
+            :meth:`~MDAnalysis.core.AtomGroup.AtomGroup.select_atoms`
+            This selection of atoms is used to calculate the RMSD between
+            different frames. Water should be excluded.
+        metric : function, optional
+            Maps two numpy arrays to a float, is positive definite and
+            symmetric, Default: metric is set to rms.rmsd().
+        cutoff : float, optional
+            Specify a given cutoff for metric values to be considered equal,
+            Default: 1EO-5
+        weights : array, optional
+            Weights to be given to coordinates for metric calculation
+        start : int, optional
+            First frame of trajectory to analyse, Default: 0
+        stop : int, optional
+            Last frame of trajectory to analyse, Default: -1
+        step : int, optional
+            Step between frames to analyse, Default: 1
+        """
+
+        self._u = u
+        traj = self._u.trajectory
+        self.atoms = self._u.select_atoms(select)
+        self._metric = metric
+        self._cutoff = cutoff
+        self._weights = weights
+        self.calculated = False
+        # remember that this must be called before referencing self.nframes
+        self._setup_frames(traj, start, stop, step)
+
+    def _prepare(self):
+        self.dist_matrix = np.zeros((self.nframes, self.nframes))
+        self._i = -1
+
+    def _single_frame(self):
+        self._i = self._i + 1
+        iframe = self._ts.frame
+        i_ref = self.atoms.positions - self.atoms.center_of_mass()
+        # diagonal entries need not be calculated due to metric(x,x) == 0 in
+        # theory, _ts not updated properly. Possible savings by setting a
+        # cutoff for significant decimal places to sparsify matrix
+        for j, ts in enumerate(self._u.trajectory[iframe:self.stop:self.step]):
+            self._ts = ts
+            j_ref = self.atoms.positions-self.atoms.center_of_mass()
+            dist = self._metric(i_ref, j_ref, weights=self._weights)
+            self.dist_matrix[self._i, j+self._i] = (dist if dist > self._cutoff
+                                                    else 0)
+            self.dist_matrix[j+self._i, self._i] = self.dist_matrix[(self._i,
+                                                                    j+self._i)]
+        self._ts = self._u.trajectory[iframe]
+
+    def _conclude(self):
+        self.calculated = True
+
+    def save(self, filename):
+        np.save(filename, self.dist_matrix)
+        logger.info("Wrote the distance-squared matrix to file %r", filename)
+
+
+class Epsilon(object):
+    """Manipulates a distance matrix by local scale parameters
+
+    Attributes
+    ----------
+    scaled_matrix : DistanceMatrix object
+        A matrix with each term divided by a local scale parameter
+    calculated : boolean
+        True after determine epsilon called
+    Methods
+    -------
+    determineEpsilon()
```

Comment:
Thanks

On Mon, Jun 27, 2016 at 12:38 PM Richard Gowers <notifications@github.com>
wrote:

> In package/MDAnalysis/analysis/diffusionmap.py
> <https://github.com/MDAnalysis/mdanalysis/pull/863#discussion_r68641700>:
>
> > +        np.save(filename, self.dist_matrix)
> > +        logger.info("Wrote the distance-squared matrix to file %r", filename)
> > +
> > +
> > +class Epsilon(object):
> > +    """Manipulates a distance matrix by local scale parameters
> > +
> > +    Attributes
> > +    ----------
> > +    scaled_matrix : DistanceMatrix object
> > +        A matrix with each term divided by a local scale parameter
> > +    calculated : boolean
> > +        True after determine epsilon called
> > +    Methods
> > +    -------
> > +    determineEpsilon()
>
> determine_epsilon
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/863/files/a12251cd2b2bc67e2f8f2163c3cd1672e420f6a2#r68641700>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKcARhgrDXilpt5uhsFGirRl3nFGx2hkks5qQCbRgaJpZM4IpYIS>
> .
>
-- 
Have a wonderful day,
John J. Detlefs